### PR TITLE
Group all event modifications into transactions

### LINKED ISSUE #1: Group similar operations into transactions.
Since on event (e,g, `on_message`) may result in the need to change multiple rows, it is more efficient and atomic if we can group it into a transaction.

The way I am thinking of doing this is to make the `sql` object stateful in storing the current transaction. All SQLAlchemy queries will go through some private method, say `_execute()`. When this method is called, if there is an active transaction, it is added. Otherwise the method will automatically query and commit it like normal via `self.db.execute()`

Then, when a client calls a `sql` object method, they will designate the creation and finish of a transaction. When signaled by another method, the transaction will attempt to commit its entire contents. If this fails, it will revert the entire transaction and print an error to the log.

It may look something like this:
```python
with sql.transaction():
    sql.add_message(message)
    sql.add_emoji(emoji)
    # etc...
```

### PR DESCRIPTION
**Pull request by ammongit**
Fixes #1 

This allows anybody with a `sql` handle to be able to do:
```python
with sql.transaction() as trans:
   sql.do_thing(trans, ...)
   sql.do_another_thing(trans, ...)
```

Also ensures that all internal `sql` methods use the correct context for this.

### CODE REVIEW

**Comment by raylu** [comment]
are you sure you want manual transaction control instead of just creating a transaction at the start of every event handler and committing/rolling back at the end automatically?

File: statbot/sql.py [hunk 1]
```diff
@@ -22,6 +22,32 @@
     'DiscordSqlHandler',
 ]
 
+class _Transaction:
+    __slots__ = (
```
Comment (raylu) [inline statbot/sql.py]: are you running out of RAM?

Comment (ammongit) [inline statbot/sql.py]: What's wrong with using `__slots__`?

**Comment by ammongit** [comment]
I don't see the point writing out all the boilerplate by hand.
It's going to do the same transaction logic for every event, so why duplicate so much code?

**Comment by raylu** [comment]
I think most people coming from using web frameworks would argue that dealing with the transaction _is_ the boilerplate and you wouldn't have it if you automatically created/cleaned up the transaction every command invocation. the version you have now is the one with the duplicated code

**Comment by ammongit** [comment]
>if you automatically created/cleaned up the transaction every command invocation

Isn't this what I'm doing by using `with`? Unless you mean adding a wrapper to all the methods that does that for me. I don't understand what you're proposing.

**Comment by raylu** [comment]
oh, I forgot you are using discord.py instead of dispatching your own events. it's a little harder to hook into that, but you can still do it by overriding `discord.Client.dispatch`https://github.com/Rapptz/discord.py/blob/7b806667cda6dc26b6a99d73473dcff5c2dd4044/discord/client.py#L316 with something like
```python
def dispatch(self, event, *args, **kwargs):
    with self.sql.transaction():
        super().dispatch(event, *args, **kwargs)
```

**Comment by ammongit** [comment]
That makes sense, but I'm inclined not to do that because I only want to start a transaction if I'm going to actually modify the database. There are many events that are triggered that are then ignored, such as members going offline, messages in servers I'm not keeping track of, etc.

**Comment by raylu** [comment]
what's the problem with starting unnecessary transactions?

**Comment by ammongit** [comment]
Because they're unnecessary..?

**Comment by ammongit** [comment]
I've changed the usage. You know must pass a transaction object with each sql method.

```python
with sql.transaction() as trans:
    sql.add_message(trans, message)
```

**Comment by raylu** [comment]
>> what's the problem with starting unnecessary transactions?

> Because they're unnecessary..?

that's... not particularly helpful. the actual question I'm asking is "given that we could provide a much simpler interface by automatically starting and committing/rolling back transactions on each event dispatch, what are the downsides of doing so?" your answer is something about some transactions being unnecessary, but I don't see a technical reason why we shouldn't do this and doing so would result in a cleaner design

**Comment by ammongit** [comment]
I don't think it would be "much simpler". Sure I could have a decorator that does the transaction for me, but that barely changes anything. Adding `@my_transaction_decorator` is one line, and adding a `with sql.transaction() as trans` is also one line (not including indentation). The complexity savings are not significant.

And there is a performance benefit. For some kinds of events, the majority of them are ignored. By adding a wrapper to the event handler as a whole, we will create many empty transactions, where if we just have the `with` scope lower, it is completely skipped over, with no transaction created at all.

**Pull request merged**